packages/store/src/StoreSwitch.sol

```diff
@@ -1,111 +1,112 @@
 // SPDX-License-Identifier: MIT
 pragma solidity >=0.8.0;
 
-import { IStore, IStoreHook } from "./IStore.sol";
+import { IStore, IStoreHook, IStoreConsumer } from "./IStore.sol";
 import { StoreCore } from "./StoreCore.sol";
 import { Schema } from "./Schema.sol";
 
 /**
  * Call IStore functions on self or msg.sender, depending on whether the call is a delegatecall or regular call.
  */
 library StoreSwitch {
-  error StoreSwitch_InvalidInsideConstructor();
+  error StoreSwitch_MissingOrInvalidStoreAddressFunction(bytes lowLevelData);
 
-  /**
-   * Detect whether the current call is a delegatecall or regular call.
-   * (The isStore method doesn't return a value to save gas, but it if exists, the call will succeed.)
-   */
-  function isDelegateCall() internal view returns (bool success) {
+  function storeAddress() internal view returns (address) {
     // Detect calls from within a constructor
     uint256 codeSize;
     assembly {
       codeSize := extcodesize(address())
     }
 
     // If the call is from within a constructor, use StoreCore to write to own storage
-    if (codeSize == 0) return true;
+    if (codeSize == 0) return address(this);
 
-    // Check whether this contract implements the IStore interface
-    try IStore(address(this)).isStore() {
-      success = true;
-    } catch {
-      success = false;
+    try IStoreConsumer(address(this)).storeAddress(msg.sender) returns (address _storeAddress) {
+      return _storeAddress;
+    } catch (bytes memory lowLevelData) {
+      // catch and rename the error, otherwise it's usually "EvmError: Revert" which is very unhelpful
+      revert StoreSwitch_MissingOrInvalidStoreAddressFunction(lowLevelData);
```

Comment:
an alternative to enforcing `IStoreConsumer` would be to fall back on `msg.sender` if it's absent. That might help with gas too.
But I like this explicit approach since it gives a nice error instead of confusing behavior down the line, making things hopefully easy to debug (or answer questions about on `mud-help`)